### ISSUE TITLE
`crucible-c` memory model issue

### ISSUE THREAD

**Issue by david-christiansen** [GaloisInc/crucible#109]
I managed to provoke the following message:
```
dtc➜Projects/crucible/crucible-c(master✗)» cabal new-run crucible-c -- ~/Projects/c-c-tests/one-sim.c                                             [16:53:31]
Up to date
[Crux] Simulating function "main"
TOP LEVEL EXCEPTION
You have encountered a bug in Crucible's implementation.
*** Please create an issue at https://github.com/GaloisInc/crucible/issues

%< --------------------------------------------------- 
  Revision:  446c208df8011f4e1c1a9dcb044a5aaa7abe0140
  Branch:    master (uncommited files present)
  Location:  MemModel.packMemValue
  Message:   Type mismatch when storing value.
             *** Expected storable type: bitvectorType 1
             *** Given crucible type: RecursiveRepr LLVM_pointer [BVRepr 1]
CallStack (from HasCallStack):
  panic, called at src/Lang/Crucible/Panic.hs:11:9 in crucible-0.4-inplace:Lang.Crucible.Panic
  panic, called at src/Lang/Crucible/LLVM/MemModel.hs:556:3 in crucible-llvm-0.2-inplace:Lang.Crucible.LLVM.MemModel
%< --------------------------------------------------- 

```

Here's (a minified version of) the code that did it:
```
#include<stdlib.h>
#include "crucible.h"

typedef struct {size_t size;} buffer;

buffer *buffer_for() {
    buffer *b = malloc(sizeof(buffer));
    b->size = 0;
    return b;
}


bool buffer_equalp(buffer *one, buffer *two) {
    // This version works:
    //return one->size == two->size;
    // This one doesn't:
    if (one->size == two->size) {
        return true;
    } else {
        return false;
    }
}

int main() {
    buffer *buf2a = buffer_for();
    check(buffer_equalp(buf2a, buf2a));
}

```

I haven't managed to make it any smaller.

**Comment by david-christiansen** [comment]
Please let me know if you can fix this quickly, or if I should go take a stab at it myself.

**Comment by robdockins** [comment]
This looks the same as #84, which is great as I've been trying to get a reproducible test case for this.  I'll look at it.

**Comment by david-christiansen** [comment]
This does indeed seem to be the same issue. Thanks!

**Comment by robdockins** [comment]
Turns out, the issue here isn't what I expected.  It's actually to do with the use of the type `bool`.  If you change the return type of `buffer_equalp` to any one of the standard C integer types, things work out.

The problem seems to be that we haven't implemented logic for handling loads and stores of values that are not multiples of 8-bits.  For reference, the LLVM manual states, regarding stores:

    When writing a value of a type like i20 with a size that is not an integral
    number of bytes, it is unspecified what happens to the extra bits that do
    not belong to the type, but they will typically be overwritten.


And on loads:

    When loading a value of a type like i20 with a size that is not an integral
    number of bytes, the result is undefined if the value was not originally written
    using a store of the same type.

**Comment by david-christiansen** [comment]
Interesting! Thanks!

**Comment by robdockins** [comment]
This issue appears to be fixed by fe51137096600ea9ab5d9b3be6962cab8d616db8
